### PR TITLE
Avoid unnecessary copy of Hadoop config in getVolumeManagerConfiguration

### PR DESCRIPTION
**Pull request by dlmarion**
User noticed BLOCKED threads in the Accumulo GarbageCollector at:

```
at org.apache.hadoop.conf.Configuration.<init>(Configuration.java:842)
at org.apache.accumulo.server.fs.VolumeManagerImpl.getVolumeManagerConfiguration(VolumeManagerImpl.java:390)
at org.apache.accumulo.server.fs.VolumeManagerImpl.getFileSystemByPath(VolumeManagerImpl.java:259)
at org.apache.accumulo.server.fs.VolumeManagerImpl.deleteRecursively(VolumeManagerImpl.java:204)
```

### CODE REVIEW

**Review by cshannon — APPROVED** [review]
This seems like a good improvement as it makes sense to not copy if those properties are not set and would avoid the lock.

Maybe this would be a follow on issue to investigate or prototype but I am also wondering if there is a way to cache a volume configuration object by INSTANCE_VOLUME_CONFIG_PREFIX. So if those properties are actually set and we do need to build a new configuration it would be nice if we could cache for re-re-use and next time and just try and look up the config by the prefix of the given fileSystemURI.

The issue with caching of course is if the configuration has changed, so I'm not sure if it's doable. We'd need a good way to know the previous cached config was invalid. Maybe we could maybe listen to the change events for the ZK config and ZooProp store to get notified there were changes?

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:399]: Would it be worthwhile to avoid creating the hadoop config copy in the case where filter on the stream drops everything?

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:399]: Yeah, so I'm trying to add that right now and running into compilation issues. It's turning the resulting map into Map<Object,Object> vs Map<String,String>. Do you have something that works?

**Comment by keith-turner** [comment]
@cshannon  re caching. If this code is called really frequently then a cache that only keeps the configuration objects for a short period of time (like a 10 second expiration time for the cache) could possibly improve performance w/o impacting correctness.  Not sure how frequently this code is hit.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:399]:
```
    Map<String,String> volumeHdfsConfigOverrides = 
        conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUME_CONFIG_PREFIX).entrySet()
            .stream().filter(e -> e.getKey().startsWith(filesystemURI + "."))
            .collect(Collectors.toUnmodifiableMap(Function.identity(), Function.identity()));
```

doesn't compile.

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:399]: I got it:

```
    Map<String,String> volumeHdfsConfigOverrides = 
        conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUME_CONFIG_PREFIX).entrySet()
            .stream().filter(e -> e.getKey().startsWith(filesystemURI + "."))
            .collect(Collectors.toUnmodifiableMap(e -> e.getKey(), e -> e.getValue()));
```

Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:399]: Could do something like

```java
    List<Entry<String, String>> matching = conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUME_CONFIG_PREFIX).entrySet()
            .stream().filter(e -> e.getKey().startsWith(filesystemURI + ".")).collect(Collectors.toList());
    
    if(!matching.isEmpty()) {
      Configuration volumeConfig = new Configuration(hadoopConf);
      matching.forEach(e->{
        String key = e.getKey().substring(filesystemURI.length() + 1);
        String value = e.getValue();
        log.info("Overriding property {} for volume {}", key, value, filesystemURI);
        volumeConfig.set(key, value);
      });
    }
```

Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:399]: This works: 

> I got it:
> 
> ```
>     Map<String,String> volumeHdfsConfigOverrides = 
>         conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUME_CONFIG_PREFIX).entrySet()
>             .stream().filter(e -> e.getKey().startsWith(filesystemURI + "."))
>             .collect(Collectors.toUnmodifiableMap(e -> e.getKey(), e -> e.getValue()));
> ```

I was just about to post it but a little different:

```
    Map<String,String> volumeHdfsConfigOverrides2 =
            conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUME_CONFIG_PREFIX).entrySet()
                    .stream().filter(e -> e.getKey().startsWith(filesystemURI + "."))
                    .collect(Collectors.toUnmodifiableMap(Entry::getKey, Entry::getValue));
```

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (matthpeterson) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:402]: This seems to be missing a {}

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: Yep, noticed and fixed in latest commit.

**Comment by dlmarion** [comment]
I think adding a cache with a short invalidation period in a follow-on PR would be a good compromise.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (cshannon) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:402]: You don't have to make this change but I noticed this could be simplified slightly with something like:

```
volumeHdfsConfigOverrides.forEach((k, v) -> {
    String key = k.substring(filesystemURI.length() + 1);
    log.info("Overriding property {}={} for volume {}", key, v, filesystemURI);
    volumeConfig.set(key, v);
});
```

Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: Other than this small nit (which you can ignore if you want as I said) the PR looks good to me. It would be nice if we could write a test for this specific `getVolumeManagerConfiguration()` method but it's private. However, it looks like the  testConfigurationOverrides() test inside of the VolumeManagerImplTest class provides coverage on the prefix usage already so I think it's ok.

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: I made your suggested change, and went ahead and moved the substring call up too in cfb3047. Yes, that test should be sufficient I think.

**Review by cshannon — APPROVED** [review]
LGTM

**Review by ctubbsii — COMMENTED** [review]
This bypasses the construction of a new configuration object when there are no overrides, but there's still a penalty if there are overrides. I think we could probably cache the result after constructing the new configuration object, and keep using it unless the overrides change. Alternatively, just cache it for a fixed period of time... and let the overrides be recomputed every 10 minutes or something like that. We just communicate that overrides will be checked no more frequently than once every 10 minutes in any particular server.

EDIT: apparently, these suggestions have already been made and I should have read before posting :smiley_cat:

**Comment by dlmarion** [comment]
> I think we could probably cache the result after constructing the new configuration object, and keep using it unless the overrides change. Alternatively, just cache it for a fixed period of time

I think the latter might be the better option. It will make the code much cleaner than trying to compare the properties in the cached configuration to properties in the current configuration.

**Comment by dlmarion** [comment]
> > I think we could probably cache the result after constructing the new configuration object, and keep using it unless the overrides change. Alternatively, just cache it for a fixed period of time
> 
> I think the latter might be the better option. It will make the code much cleaner than trying to compare the properties in the cached configuration to properties in the current configuration.

So, I went back and forth on this in e5a8b7e. I ended up being concerned about caching for a fixed-period. If the period is long, then that may introduce issues. If the period is short, then you end up recreating the Configuration object and creating a synchronization point when you might not need to. So, I opted to cache for a longer period of time and recreate the Configuration if the overrides changed before that.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (cshannon) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:464]: I think this can be deleted.

**Review by cshannon — CHANGES_REQUESTED** [review]
This should work but I was going over it and think it would just be simpler/easier if instead of the key being the URI we just made the key the volumeHdfsConfigOverrides map. We wouldn't need to worry about having the extra object as the key or having to use the boolean to detect new entries.

The issue with that change of course is if the properties change you end up with a stale entry in the Cache that won't ever be used anymore. However, I think this is probably fine. You still have a 6 hour time out which would evict stale entries if the properties changed. Also, as a precaution we could set a max number of entries on the cache just to prevent blowing heap in some weird edge case or bug where properties kept changing and if too many entires went into the cache it would evict based on least recently used with the tinyLRU algorithm it uses.

**Comment by cshannon** [comment]
Also, as a side note, if the map is used as the key we need to make sure the map implementation properly implements hashcode/equals to do a deep comparison of the entries. This is the same for the current PR though, as the current version is doing a comparsion of the cached map with a new map that is created from the prefix configs on each call to getVolumeManagerConfiguration().

The change here is using an ImmutableMap from ImmutableCollections so I think this already does a proper hashcodes/equals deep comparsion but need to double check.

**Comment by keith-turner** [comment]
Thinking the following two things taken together could simplify the caching and avoid the check to see if things in the cache changed the code is currently doing. A short timeout plus the fact that the props are not even expected to change makes me think we could avoid checking if things changed.  The short timeout means if the assumption about instance props not changing is no longer true int he future that the code would be  still be resilient to changes.

 * instance props are not ZK mutable, so unless Accumulo periodically reloads the site file these props should be stable for the lifetime of the process.
 * I suspect the cache timeout could be much shorter than 6 hours.  A timeout of a few seconds should give a really nice amortized cost for the case where there are many thousands of accesses per second.  For example if the timeout is 10 seconds and its accessed 10K times per a second on average then it would only be recomputed on every 100K access on averge which I suspect will not impact performance noticeably

**Comment by keith-turner** [comment]
Also thinking if we shorten the timeout and assume instance props do not change that using the URI for the cache key like the code is currently doing should be ok.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:439]: This caching something derived from a method parameter, the hadoopConfig. Need to account for different configs somehow or refactor the code to not pass in a hadoopConfig.

**Comment by dlmarion** [comment]
> instance props are not ZK mutable

I forgot about this. It would be great if this was part of the Property declarations. If this is the case, then:

  * we just need to create a new Configuration for each volume that has overrides once, for the duration of the process.
  * the filesystem URI as the key should suffice, as the override map will never change.
  * I don't need to check for property changes in the method
  
I would advocate for a long ache timeout if things are not expected to change. That will reduce the probability of a synchronization point between threads.

**Comment by EdColeman** [comment]
Could the timeout be based on the gc run frequency?  Say 1/4 or 1/2 of the `GC_CYCLE_DELAY` - that should enforce that on each run the config is checked / cached and should not need to be renewed frequently.

**Comment by dlmarion** [comment]
> Could the timeout be based on the gc run frequency? Say 1/4 or 1/2 of the `GC_CYCLE_DELAY` - that should enforce that on each run the config is checked / cached and should not need to be renewed frequently.

`VolumeManager.get` and `VolumeManager.getFileSystemByPath` are called in many other places, it's not just GC.

**Comment by dlmarion** [comment]
Confirmed that property will not change, removed logic for handling that in c8a5715. That made the implementation much simpler.

**Review by cshannon — APPROVED** [review]
LGTM, it's definitely much simpler now since we don't need to handle the case of changing properties.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:410]: This is a bit hacky, but it seems like it could work to support different hadoopConfig instances being passed in.  When the exact same object is passed in should use the cached entry.

```suggestion
    /**
     * Hadoop does not implement hashCode for the configuration class, so the following gets the
     *  objects hashcode.  The concept is to make the cached object have 1:1 correspondence with
     *  the hadoop config object from which it was derived.
     */
     String key = filesystemURI+hadoopConf.hashCode();

      return HDFS_CONFIGS_FOR_VOLUME.get(key, (fs) -> {
```

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: I think there is only one Hadoop configuration object in use. I didn't track this down though and instead just implemented your suggestion in 1b2fb52.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:396]: Is it ok to do the following and void the computation of volumeHdfsConfigOverrides on each call?

```java
    final String key = hadoopConf.hashCode() + filesystemURI;
    return HDFS_CONFIGS_FOR_VOLUME.get(key, (fs) -> {

      Map<String,String> volumeHdfsConfigOverrides =
              conf.getAllPropertiesWithPrefixStripped(Property.INSTANCE_VOLUME_CONFIG_PREFIX).entrySet()
                      .stream().filter(e -> e.getKey().startsWith(filesystemURI + ".")).collect(
                              Collectors.toUnmodifiableMap(e -> e.getKey().substring(filesystemURI.length() + 1),
                                      Entry::getValue));
      if (volumeHdfsConfigOverrides.isEmpty()) {
        return hadoopConf;
      }else {
        log.debug("Caching a new configuration for volume: {} with overrides: {}", filesystemURI,
                volumeHdfsConfigOverrides);
        Configuration volumeConfig = new Configuration(hadoopConf);
        volumeHdfsConfigOverrides.forEach((k, v) -> {
          log.info("Overriding property {}={} for volume {}", k, v, filesystemURI);
          volumeConfig.set(k, v);
        });
        return new Configuration(volumeConfig);
      }
    });
```

**Review by keith-turner — APPROVED** [review]

**Comment by ctubbsii** [comment]
I'm still reviewing this and had some suggestions, but my review keeps changing, because the updates are causing me to refresh. Just letting you know, in case you wanted to wait until I wrap up my review. It shouldn't be long.

**Comment by dlmarion** [comment]
I'll wait.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:396]: My only concern with putting more into the `Cache.get` call is this in the javadoc:
```
   * Some attempted update operations on this cache by other threads may be blocked while the
   * computation is in progress, so the computation should be short and simple, and must not attempt
   * to update any other mappings of this cache.
```

Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:396]: Could consider this change later if the computation of volumeHdfsConfigOverrides actually shows up in any profiling.

**Review by ctubbsii — APPROVED** [review]
Talked with Dave off ticket and pushed my code review suggestions up as a commit on the PR. I think it's good to go.

**Comment by dlmarion** [comment]
LGTM. I'll merge on Monday assuming no other suggestions.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: I included this suggestion in my code review commit 8f1ee93431999e673a878be134f947270eb0dd17

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java:415]: This logging seems to duplicate the information that will be logged in the later log.info at the debug level.  Seems like it could be removed

**Comment by cshannon** [comment]
I saw the comment about the performance of blocking the get() when the new cache entry is computed due to the performance of building the map inside of that function since it will be executed under a lock.

If this becomes an issue of contention we can solve this by using an async cache. https://github.com/ben-manes/caffeine/wiki/Population#asynchronous-manual

The get() would return a future and it wouldn't block access by other threads while computing. I think the main downside would be maybe creating the same object more than once if multiple threads hit it but once it's cached then won't be an issue.

**Comment by ben-manes** [comment]
@cshannon That's a good approach for decoupling the computation from the map. The mapping to create the future is performed atomically, so you won't observe a get-load-put race when multiple threads try to memoize. The main drawback is that the future's work is dispatched onto an executor, so if simply blocking then a bit wasteful until virtual threads arrive. That can be mitigated using JCiP style memoizer (putIfAbsent with a future), but is more manual code. If not a problem, then you can use the `synchronous()` view to trivially swap over while keeping the otherwise code. 

A simpler trick is to increase the `initialCapacity` of the underlying ConcurrentHashMap. It determines the number of lock stripes by the table size, so a small map has a higher chance of hash collisions by concurrent writes. A larger table does not waste much and allocates more locks, so the likelihood reduces.

The [FAQ](https://github.com/ben-manes/caffeine/wiki/Faq) touches on some of these topics.

**Comment by cshannon** [comment]
> @cshannon That's a good approach for decoupling the computation from the map. The mapping to create the future is performed atomically, so you won't observe a get-load-put race when multiple threads try to memoize. The main drawback is that the future's work is dispatched onto an executor, so if simply blocking then a bit wasteful until virtual threads arrive. That can be mitigated using JCiP style memoizer (putIfAbsent with a future), but is more manual code. If not a problem, then you can use the `synchronous()` view to trivially swap over while keeping the otherwise code.
> 
> A simpler trick is to increase the `initialCapacity` of the underlying ConcurrentHashMap. It determines the number of lock stripes by the table size, so a small map has a higher chance of hash collisions by concurrent writes. A larger table does not waste much and allocates more locks, so the likelihood reduces. The expectation is that the map's write write likely correlates to its size, as locks are dynamically added when the table grows, but if that is not the case then setting the initial capacity is a pretty simple workaround.
> 
> The [FAQ](https://github.com/ben-manes/caffeine/wiki/Faq) touches on some of these topics.

@ben-manes - that is very helpful information and some nice tips, thanks for taking the time to respond explain in more detail about how the async cache handles the computation. It's always great to hear from the creator of a library when you have a question about how something works. :)

**Pull request merged**
